src/Front/ApplePay.php

```diff
@@ -435,4 +436,25 @@ public function applepay_cancel_order()
 
 	}
 
+	public function is_shipping_required() {
+		$cart = WC()->cart->get_cart();
+
+		$required = true;
+		foreach ( $cart as $cart_item ) {
+			if ( ! empty( $cart_item['product_id'] ) ) {
+				$product = wc_get_product( $cart_item['product_id'] );
+
+				//not required if it enters here
+				if ( ! $product->is_virtual() && ! $product->is_downloadable() ) {
+					return wp_send_json(true);
+				}
+
+				$required = false;
+			}
+		}
```

Comment:
I still have the same comment, this can be simplified and optimized to only have 1 return point 
```suggestion
$required = false;
	foreach ( $cart as $cart_item ) {
		if ( ! empty( $cart_item['product_id'] ) ) {
			$product = wc_get_product( $cart_item['product_id'] );
			//not required if it enters here
			if ( ! $product->is_virtual() && ! $product->is_downloadable() ) {
				$required = true;
				break;
			}
		}
	}

	return wp_send_json($required);
```

Reply:
all this has been deleted because I changed the solution